### PR TITLE
unix: support SO_REUSEPORT with load balancing for UDP

### PR DESCRIPTION
**Pull request by panjf2000**
UDP version of #4407

### CODE REVIEW

**Comment by panjf2000** [comment]
This [CI failure](https://github.com/libuv/libuv/actions/runs/9207064232/job/25326279524?pr=4419) should be fixed by #4417

**Comment by panjf2000** [comment]
Kindly ping @saghul @santigimeno

**Comment by saghul** [comment]
Merged the test update, can you pl rebase?

**Comment by panjf2000** [comment]
> Merged the test update, can you pl rebase?

Done!

File: test/test-udp-reuseport.c [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+  r = uv_mutex_init(&mutex);
+
+  r = uv_sem_init(&semaphore, 0);
+  ASSERT_OK(r);
+
+  main_loop = uv_default_loop();
+  ASSERT_NOT_NULL(main_loop);
+
+  /* Create receiving socket per event loop. */
+  thread_loop1 = uv_loop_new();
```
Comment (santigimeno) [inline test/test-udp-reuseport.c:188]: Can we avoid using `uv_loop_new()` and `uv_loop_delete()` as they're deprecated

File: test/test-udp-reuseport.c [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+  ASSERT_OK(r);
+}
+
+TEST_IMPL(udp_reuseport) {
+  struct sockaddr_in addr;
+  uv_buf_t buf;
+  int r;
+  int i;
+
+  r = uv_mutex_init(&mutex);
```
Comment (santigimeno) [inline test/test-udp-reuseport.c:179]:
```suggestion
ASSERT_OK(uv_mutex_init(&mutex));
```

**Comment by panjf2000** [comment]
A test failure arose:
```
not ok 51 - fs_event_error_reporting
# exit code 6
# Output from process `fs_event_error_reporting`:
# Assertion failed in /Users/runner/work/libuv/libuv/test/test-fs-event.c on line 1115: `uv_loop_init(loop) == 0` (-24 == 0)
```
, which should have nothing to do with this PR (hopefully (⊙ˍ⊙))

File: test/test-udp-reuseport.c [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+  ASSERT_NOT_NULL(thread_loop1);
+  bind_socket_and_prepare_recv(thread_loop1, &thread_handle1);
+  uv_timer_init(thread_loop1, &thread_timer_handle1);
+  uv_timer_start(&thread_timer_handle1, ticktack, 0, 10);
+
+  thread_loop2 = uv_loop_new();
+  ASSERT_NOT_NULL(thread_loop2);
+  bind_socket_and_prepare_recv(thread_loop2, &thread_handle2);
+  uv_timer_init(thread_loop2, &thread_timer_handle2);
+  uv_timer_start(&thread_timer_handle2, ticktack, 0, 10);
```
Comment (santigimeno) [inline test/test-udp-reuseport.c:198]: Personally I'd call the `bind_socket_and_prepare_recv()` methods from the actual thread the loop is running on.

**Review by santigimeno — APPROVED** [review]
LGTM. Left a couple of comments.

File: test/test-udp-reuseport.c (hunk 1 above)
Comment (panjf2000) [reply test/test-udp-reuseport.c]: Done! Also updated the `test-tcp-reuseport.c`

File: test/test-udp-reuseport.c (hunk 2 above)
Comment (panjf2000) [reply test/test-udp-reuseport.c:179]: Ditto.

File: test/test-udp-reuseport.c (hunk 3 above)
Comment (panjf2000) [reply test/test-udp-reuseport.c]: I guess we can do that. But if we do that, it seems that we may also want to move `uv_timer_init` and `uv_timer_start` there as well. Is that an actual benefit or just a preference?

Comment (santigimeno) [reply test/test-udp-reuseport.c]: Yes those too. It's a personal preference in the sense that most of libuv methods aren't thread-safe so just to be sure they're always called from the loop thread it's a good idea in order to avoid problems. Of course in this case it's not an issue as the threads haven't been started yet. Feel free to dismiss the comment though.

Comment (panjf2000) [reply test/test-udp-reuseport.c]: Yes, both ways should have the same effect. But indeed your suggestion will make the code more feasible for each thread to initialize their own data. I'd like to go with that for `tcp_reuseport` and `udp_reuseport`.

**Comment by panjf2000** [comment]
Some comments on this? @saghul

File: docs/src/udp.rst [hunk 4]
```diff
@@ -186,11 +198,17 @@ API
...
-        ``UV_UDP_IPV6ONLY``, ``UV_UDP_REUSEADDR``, and ``UV_UDP_RECVERR``
-        are supported.
+        ``UV_UDP_IPV6ONLY``, ``UV_UDP_REUSEADDR``, ``UV_UDP_REUSEPORT``,
+        and ``UV_UDP_RECVERR`` are supported.
 
     :returns: 0 on success, or an error code < 0 on failure.
 
+    .. versionchanged:: 1.49.0 added the ``UV_UDP_REUSEPORT`` flag.
+
+    .. note::
```
Comment (saghul) [inline docs/src/udp.rst:208]: Let's document what happens if the flag is set in a system which doesn't support it.

File: include/uv.h [hunk 5]
```diff
@@ -682,6 +682,18 @@ enum uv_udp_flags {
    * This flag is no-op on platforms other than Linux.
    */
   UV_UDP_LINUX_RECVERR = 32,
+  /*
+   * Indicates if SO_REUSEPORT will be set when binding the handle.
```
Comment (saghul) [inline include/uv.h:686]: Please update the comment in `UV_UDP_REUSEADDR` since it does mention REUSEPORT on some systems.

File: src/unix/core.c [hunk 6]
```diff
@@ -1910,3 +1910,48 @@ unsigned int uv_available_parallelism(void) {
   return (unsigned) rc;
 #endif  /* __linux__ */
 }
+
+int uv__sock_reuseport(int fd) {
```
Comment (saghul) [inline src/unix/core.c:1914]: We already have the internal `uv__set_reuse`. Perhaps it would be a good idea to rename that one to `uv__sock_reuseaddr` for clarity now.

File: src/unix/udp.c [hunk 7]
```diff
@@ -532,6 +533,12 @@ int uv__udp_bind(uv_udp_t* handle,
       return err;
   }
 
+  if (flags & UV_UDP_REUSEPORT) {
```
Comment (saghul) [inline src/unix/udp.c:536]: What is the effect of setting both reuseaddr and reuseport? Is setting the latter enough for the load sharing case? Let's have that documented please.

File: include/uv.h (hunk 5 above)
Comment (panjf2000) [reply include/uv.h:686]: Update the comment with what? About what platforms **will not** be set with `SO_REUSEPORT` when specifying `UV_UDP_REUSEADDR`?

File: src/unix/core.c (hunk 6 above)
Comment (panjf2000) [reply src/unix/core.c:1914]: `uv__sock_reuseaddr` might be a bit confusing given that this function preferentially sets `SO_REUSEPORT` instead of `SO_REUSEADDR`.

File: src/unix/udp.c (hunk 7 above)
Comment (panjf2000) [reply src/unix/udp.c:536]: For platforms where `SO_REUSEPORT`s have load balancing, setting both options is allowed and `SO_REUSEPORT` will always override the behavior of `SO_REUSEADDR`.

For platforms where `SO_REUSEPORT`s don't have load balancing, setting both options will fail: `uv__set_reuse()` will succeed but `uv__sock_reuseport` will return `UV_ENOTSUP`.

File: docs/src/udp.rst (hunk 4 above)
Comment (panjf2000) [reply docs/src/udp.rst:208]: Consider it done.

File: include/uv.h (hunk 5 above)
Comment (saghul) [reply include/uv.h:686]: Yes please.

File: src/unix/core.c (hunk 6 above)
Comment (saghul) [reply src/unix/core.c:1914]: Got any other ideas?

Comment (panjf2000) [reply src/unix/core.c:1914]: As confusing as `uv__sock_reuseaddr` might be, I don't think we have another choice. So I'd say we rename it `uv__sock_reuseaddr` but add some comments to explain what will really happen in that function. What do you think?

File: include/uv.h (hunk 5 above)
Comment (panjf2000) [reply include/uv.h:689]: Done.

File: src/unix/udp.c (hunk 7 above)
Comment (panjf2000) [reply src/unix/udp.c:539]: Done. Added documentations along with new tests.

**Review by saghul — APPROVED** [review]
👏

**Comment by panjf2000** [comment]
It seems like an irrelevant CI failure:
```
not ok 321 - tcp_oob
# timeout
# Output from process `tcp_oob`: (no output)
```

Could you rerun the failed runner manually? @saghul

**Comment by saghul** [comment]
> It seems like an irrelevant CI failure:
> 
> ```
> not ok 321 - tcp_oob
> # timeout
> # Output from process `tcp_oob`: (no output)
> ```
> 
> Could you rerun the failed runner manually? @saghul

Done.

**Comment by panjf2000** [comment]
Great! All checks have passed!
Can we get this PR merged now? @saghul

**Comment by panjf2000** [comment]
Ping @libuv/collaborators

**Comment by panjf2000** [comment]
@santigimeno Could you take a quick look at the latest code changes? Thanks!

File: test/test-udp-reuseport.c [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+static uv_mutex_t mutex;
+static unsigned int received;
+
+static unsigned int thread_loop1_recv;
+static unsigned int thread_loop2_recv;
+static unsigned int sent;
+
+static uv_loop_t* main_loop;
+static uv_loop_t* thread_loop1;
+static uv_loop_t* thread_loop2;
```
Comment (santigimeno) [inline test/test-udp-reuseport.c:91]:
```suggestion
static uv_loop_t thread_loop1;
static uv_loop_t thread_loop2;
```
Nit: don't malloc the loops would make the code cleaner

File: src/unix/core.c [hunk 9]
```diff
@@ -1677,7 +1677,7 @@ int uv_thread_setpriority(uv_thread_t tid, int priority) {
     param.sched_priority = prio;
     r = pthread_setschedparam(tid, policy, &param);
     if (r != 0)
-      return UV__ERR(errno);  
+      return UV__ERR(errno);
```
Comment (santigimeno) [inline src/unix/core.c:1680]: Nit: I would not include the unrelated style changes so git blame is more useful

**Review by santigimeno — APPROVED** [review]
LGTM with a couple of non-mandatory nits. Sorry it took this long getting to review this: was very busy. Thanks

**Review by santigimeno — APPROVED** [review]
